### DISCUSSION TITLE
Socket closed

### DISCUSSION THREAD

**Discussion by cyrus937** [libp2p/rust-libp2p#4979] [Q&A] [answered]
I've developed a module that lets me create a peer-to-peer network based on **rust-libp2p**. I'm currently trying to set up a binding between this code and my python code, but I'm getting this error from mdns : 

```
2023-12-04T03:29:30.498895Z ERROR libp2p_mdns:: behaviour: if watch returned an error: rtnetlink socket closed 
2023-12-04T03:29:30.498899Z ERROR netlink_proto::framed: failed to read from netlink socket: Custom { kind: Other, error : "A Tokio 1.x context was found, but it is being shutdown." }
```

Do you have any idea how to solve it?

**Comment by thomaseizinger** [comment]
Can you post an example to reproduce it? My guess is you are dropping the runtime as you exit the FFI function.

**Reply by cyrus937** [reply]
I've written two functions, one to create the node and another to listen for events:

`async fn create_node(
    address: Option<&str>,
    port: Option<u32>,
    tcp: Option<bool>,
    udp: Option<bool>,
    msg_topic: Option<&str>,
) -> Result<Node, Box<dyn Error>> {
    let addr: &str = match address {
        Some(val) => val,
        None => "0.0.0.0",
    };

    let po = match port {
        Some(val) => val,
        None => 0,
    };

    let _ = tracing_subscriber::fmt()
        .with_env_filter(EnvFilter::from_default_env())
        .try_init();

    let mut swarm = libp2p::SwarmBuilder::with_new_identity()
        .with_tokio()
        .with_tcp(
            tcp::Config::default().port_reuse(true).nodelay(true),
            noise::Config::new,
            yamux::Config::default,
        )
        .unwrap()
        .with_quic()
        .with_behaviour(|key| {
            let message_id_fn = |message: &gossipsub::Message| {
                let mut s = DefaultHasher::new();
                message.data.hash(&mut s);
                gossipsub::MessageId::from(s.finish().to_string())
            };

            let gossipsub_config = gossipsub::ConfigBuilder::default()
                .heartbeat_interval(Duration::from_secs(10)) 
                .validation_mode(gossipsub::ValidationMode::Strict) 
                .message_id_fn(message_id_fn) 
                .build()
                .map_err(|msg| io::Error::new(io::ErrorKind::Other, msg))?;

            let gossipsub = gossipsub::Behaviour::new(
                gossipsub::MessageAuthenticity::Signed(key.clone()),
                gossipsub_config,
            )?;

            let mdns =
                mdns::tokio::Behaviour::new(mdns::Config::default(), key.public().to_peer_id())?;
            Ok(MyBehaviour { gossipsub, mdns })
        })
        .unwrap()
        .with_swarm_config(|c| c.with_idle_connection_timeout(Duration::from_secs(60)))
        .build();

    let topic = gossipsub::IdentTopic::new(msg_topic.unwrap());
    
    swarm.behaviour_mut().gossipsub.subscribe(&topic).unwrap();

    if tcp.unwrap() {
        let tcp_address = format!("/ip4/{}/tcp/{}", addr, po);
        swarm
            .listen_on(tcp_address.to_owned().parse().unwrap())
            .unwrap();
    }
    if udp.unwrap() {
        let udp_address = format!("/ip4/{}/udp/{}/quic-v1", addr, po);
        swarm
            .listen_on(udp_address.to_owned().parse().unwrap())
            .unwrap();
    } else if !tcp.unwrap() {
        let tcp_address = format!("/ip4/{}/tcp/{}", addr, po);
        swarm
            .listen_on(tcp_address.to_owned().parse().unwrap())
            .unwrap();
    }

    Ok(Node {
        id: *swarm.local_peer_id(),
        node_listening: HashSet::new(),
        peers_id: HashSet::new(),
        swarm: swarm,
        topic,
    })
}`

`async fn run(node: &mut Node) {
    let mut stdin = io::BufReader::new(io::stdin()).lines();

    loop {
        let evt = {
            tokio::select! {
                line = stdin.next_line() => Some(EventType::Input(line.expect("can get line").expect("can read line from stdin"))),
                event = node.swarm.select_next_some() => match event {
                    SwarmEvent::Behaviour(MyBehaviourEvent::Gossipsub(gossipsub::Event::Message {
                                    propagation_source: peer_id,
                                    message_id: id,
                                    message,
                                })) => {println!(
                                        "Got message: '{}' with id: {id} from peer: {peer_id}",
                                        String::from_utf8_lossy(&message.data),
                                    );
                                None
                            },
                                SwarmEvent::Behaviour(MyBehaviourEvent::Mdns(mdns::Event::Discovered(list))) => {
                                    for (peer_id, _multiaddr) in list {
                                        // println!("mDNS discovered a new peer: {peer_id}");
                                        node.swarm.behaviour_mut().gossipsub.add_explicit_peer(&peer_id);
                                    }; None
                                },
                                SwarmEvent::Behaviour(MyBehaviourEvent::Mdns(mdns::Event::Expired(list))) => {
                                    for (peer_id, _multiaddr) in list {
                                        println!("mDNS discover peer has expired: {peer_id}");
                                        node.swarm.behaviour_mut().gossipsub.remove_explicit_peer(&peer_id);
                                    }; None
                                },
                    _ => {None}
                },
            }
        };

        if let Some(event) = evt {
            match event {
                EventType::Response(resp) => {
                    println!("{}", resp)
                }
                EventType::Input(line) => match line.as_str() {
                    "ls p" => async_handle_list_peers(node).await,
                    "ls l" => node.handle_list_listening_node(),
                    cmd if cmd.starts_with("send ") => node.send_message(cmd),
                    cmd if cmd.starts_with("connected ") => node.is_connected_to(cmd),
                    cmd if cmd.starts_with("connect ") => node.connect_to(cmd),
                    cmd if cmd.starts_with("disconnect ") => node.disconnect_to(cmd),
                    "stop" => break,
                    _ => println!("unknown command"),
                },
            }
        }
    }
}`

My python code is as follows

`import libp2p_node


if __name__ == '__main__':
    node = libp2p_node.Node("0.0.0.0", 0, True, True, "test-net")
    node.run()`

**Reply by thomaseizinger** [reply]
Are you starting a tokio runtime around that 2nd function?

**Reply by cyrus937** [reply]
Yes, in my node's run method implementation

`fn run(&mut self) -> PyResult<()> {
        let rt = tokio::runtime::Runtime::new().unwrap();
        rt.block_on(async {
            run(self).await;
        });
        Ok(())
    }`

**Reply by thomaseizinger** [reply]
It is hard to say with individual code snippets, can you post a repository that reproduces the problem and is as minimal as possible?

**Reply by thomaseizinger** [reply]
I've not done interop with python before. Perhaps try some non-libp2p async stuff with Python and see if you can find something there! :)

**Reply by cyrus937** [reply]
Here is the link to the public repository : 

https://github.com/cyrus937/libp2p-node

**Reply by thomaseizinger** [reply, marked as answer]
Thanks for sharing the code. I'd suggest to reconsider the design a bit that you've chosen.

For one, you are directly interacting with stdio from _within_ the Rust code. It would be better to employ message passing via a channel. See https://github.com/libp2p/rust-libp2p/blob/master/examples/file-sharing/src/network.rs for some inspiration.

Secondly, it is **really** important that you continuously, without interruption call `Swarm::next()` and handle each event with **minimal** latency. The `Swarm` is essentially a giant `Future` and **will not** make progress if you stop polling it. If you follow the above pattern of an eventloop with message passing, that is solved for you.

In regards to your FFI design, I'd suggest to start the tokio runtime in there and have an endless loop in which you poll the swarm for events using `Swarm::next`. In other words _don't_ return from your FFI function. That will de-allocate the runtime again and stop any pending tasks and connections immediately. That is likely why you are seeing sockets being closed. Connections run on independent tasks in parallel in the background but if you drop the `Runtime`, they have nowhere to execute.

**Reply by cyrus937** [reply]
Thank you for your feedback, and sorry for the delay of my feedback.

I made the changes as you suggested, using the tokio runtime block_on, but I still get the same error with the sockets closing.

Sorry if I don't quite understand what you're saying, but I'm new to using Rust, more specifically the rust-libp2p library.

repo: https://github.com/cyrus937/libp2p-node

```rust
fn run(&mut self) {
        let handle = tokio::runtime::Runtime::new().expect("Failed to create Tokio runtime");
        handle.block_on(
            async {

                let command_arc_clone = Arc::clone(&self.node.command_receiver);
                let mut command_guard = command_arc_clone.lock().await;
                let command_receiver = &mut *command_guard;

                loop {
                    tokio::select! {
                        event = self.node.swarm.next() => handle_event(event.expect("Swarm stream to be infinite."), &mut self.node.swarm).await,
                        command = command_receiver.next() => match command {
                            Some(c) => {
                                handle_command(c, &mut self.node.swarm).await},
                            // Command channel closed, thus shutting down the network event loop.
                            None=>  {},
                        },
                    };
                }
            }
        );
    }
```

**Reply by thomaseizinger** [reply]
> Sorry if I don't quite understand what you're saying, but I'm new to using Rust, more specifically the rust-libp2p library.

No worries! There are quite a few resources out there to learn from! You are touching on many subjects at the same time. I'd recommend to explore them in isolation so you can get a better understanding of them (in no particular order):

- Try to use `rust-libp2p` standalone
- Use Rust FFI with Python
- Learn Rust async

One thing to note above is that you shouldn't block the event loop on handling the command, i.e. `handle_command` will stop the swarm from making progress while it is handling the command. The fact that you are passing the `Swarm` in means that you are likely deadlocking there (semantically).

**Reply by cyrus937** [reply]
Thanks !!!
